File: Tests/KiboCoreTests/RunSplitterTests.swift
import XCTest
@testable import KiboCore

final class RunSplitterTests: XCTestCase {

    /// Compact spelling of an expected split, so a case reads as close to the input as possible.
    private func assertSplit(_ input: String,
                             into expected: [(Run.Script, String)],
                             file: StaticString = #filePath, line: UInt = #line) {
        let runs = RunSplitter.split(input)
        XCTAssertEqual(runs.map(\.text), expected.map(\.1), "texts", file: file, line: line)
        XCTAssertEqual(runs.map(\.script), expected.map(\.0), "scripts", file: file, line: line)
    }

    func test_empty_input_produces_no_runs() {
        XCTAssertTrue(RunSplitter.split("").isEmpty)
    }

    func test_single_script_input_is_one_run() {
        assertSplit("hello", into: [(.latin, "hello")])
        assertSplit("สวัสดี", into: [(.thai, "สวัสดี")])
    }

    /// Whitespace is neutral, which is what makes it a run boundary — two Latin words separated
    /// by a space are judged separately rather than as one blob.
    func test_space_separates_runs_and_is_preserved() {
        assertSplit("hello world", into: [
            (.latin, "hello"), (.neutral, " "), (.latin, "world"),
        ])
    }

    func test_the_worked_example_from_the_spec() {
        assertSplit("l;ylfu ้ำสสน ครับ 2024 :)", into: [
            (.latin, "l;ylfu"), (.neutral, " "),
            (.thai, "้ำสสน"), (.neutral, " "),
            (.thai, "ครับ"), (.neutral, " "),
            (.latin, "2024"), (.neutral, " "),
            (.latin, ":)"),
        ])
    }

    /// Punctuation stays inside a Latin run. It has to: `;` is the `ว` key, so dropping it from
    /// the run would break `l;ylfu` → `สวัสดี`.
    func test_ascii_punctuation_belongs_to_the_latin_run() {
        assertSplit("l;ylfu", into: [(.latin, "l;ylfu")])
        assertSplit("don't", into: [(.latin, "don't")])
    }

    func test_script_change_without_whitespace_still_splits() {
        assertSplit("helloสวัสดี", into: [(.latin, "hello"), (.thai, "สวัสดี")])
    }

    /// Anything that is neither Thai nor printable ASCII is neutral and passes through: emoji,
    /// newlines, tabs, and other scripts.
    func test_non_thai_non_ascii_is_neutral() {
        assertSplit("hi🐈there", into: [
            (.latin, "hi"), (.neutral, "🐈"), (.latin, "there"),
        ])
        assertSplit("a\n\tb", into: [
            (.latin, "a"), (.neutral, "\n\t"), (.latin, "b"),
        ])
        assertSplit("café", into: [(.latin, "caf"), (.neutral, "é")])
        assertSplit("日本", into: [(.neutral, "日本")])
    }

    /// Adjacent neutral scalars coalesce rather than producing a run each.
    func test_consecutive_neutrals_form_one_run() {
        assertSplit("a   b", into: [(.latin, "a"), (.neutral, "   "), (.latin, "b")])
    }

    /// Splitting must not lose or reorder anything — the runs always rejoin to the input.
    func test_runs_always_rejoin_to_the_input() {
        let inputs = [
            "", "hello", "สวัสดี", "l;ylfu ้ำสสน ครับ 2024 :)", "hi🐈there",
            "a\n\tb", "café", "日本", "  ", "ๆๆๆ!!!",
        ]
        for input in inputs {
            XCTAssertEqual(RunSplitter.split(input).map(\.text).joined(), input, "input: \(input)")
        }
    }

    /// Thai combining marks must stay attached to the run, not be split off as neutral.
    func test_thai_combining_marks_stay_in_the_thai_run() {
        let runs = RunSplitter.split("ครับ")
        XCTAssertEqual(runs.count, 1)
        XCTAssertEqual(runs.first?.text.unicodeScalars.count, 4)
    }
}

File: Windows/Kibo.Core.Tests/RunSplitterTests.cs
namespace Kibo.Core.Tests;

public class RunSplitterTests
{
    /// Compact spelling of an expected split, so a case reads as close to the input as possible.
    private static void AssertSplit(string input, params (Script Script, string Text)[] expected)
    {
        var runs = RunSplitter.Split(input);
        Assert.Equal(expected.Select(e => e.Text), runs.Select(r => r.Text));
        Assert.Equal(expected.Select(e => e.Script), runs.Select(r => r.Script));
    }

    [Fact]
    public void Empty_input_produces_no_runs()
    {
        Assert.Empty(RunSplitter.Split(""));
    }

    [Fact]
    public void Single_script_input_is_one_run()
    {
        AssertSplit("hello", (Script.Latin, "hello"));
        AssertSplit("สวัสดี", (Script.Thai, "สวัสดี"));
    }

    /// Whitespace is neutral, which is what makes it a run boundary — two Latin words separated
    /// by a space are judged separately rather than as one blob.
    [Fact]
    public void Space_separates_runs_and_is_preserved()
    {
        AssertSplit("hello world", (Script.Latin, "hello"), (Script.Neutral, " "), (Script.Latin, "world"));
    }

    [Fact]
    public void The_worked_example_from_the_spec()
    {
        AssertSplit("l;ylfu ้ำสสน ครับ 2024 :)",
            (Script.Latin, "l;ylfu"), (Script.Neutral, " "),
            (Script.Thai, "้ำสสน"), (Script.Neutral, " "),
            (Script.Thai, "ครับ"), (Script.Neutral, " "),
            (Script.Latin, "2024"), (Script.Neutral, " "),
            (Script.Latin, ":)"));
    }

    /// Punctuation stays inside a Latin run. It has to: `;` is the `ว` key, so dropping it from
    /// the run would break `l;ylfu` → `สวัสดี`.
    [Fact]
    public void Ascii_punctuation_belongs_to_the_latin_run()
    {
        AssertSplit("l;ylfu", (Script.Latin, "l;ylfu"));
        AssertSplit("don't", (Script.Latin, "don't"));
    }

    [Fact]
    public void Script_change_without_whitespace_still_splits()
    {
        AssertSplit("helloสวัสดี", (Script.Latin, "hello"), (Script.Thai, "สวัสดี"));
    }

    /// Anything that is neither Thai nor printable ASCII is neutral and passes through: emoji,
    /// newlines, tabs, and other scripts.
    [Fact]
    public void Non_thai_non_ascii_is_neutral()
    {
        AssertSplit("hi🐈there", (Script.Latin, "hi"), (Script.Neutral, "🐈"), (Script.Latin, "there"));
        AssertSplit("a\n\tb", (Script.Latin, "a"), (Script.Neutral, "\n\t"), (Script.Latin, "b"));
        AssertSplit("café", (Script.Latin, "caf"), (Script.Neutral, "é"));
        AssertSplit("日本", (Script.Neutral, "日本"));
    }

    /// Adjacent neutral scalars coalesce rather than producing a run each.
    [Fact]
    public void Consecutive_neutrals_form_one_run()
    {
        AssertSplit("a   b", (Script.Latin, "a"), (Script.Neutral, "   "), (Script.Latin, "b"));
    }

    /// Splitting must not lose or reorder anything — the runs always rejoin to the input.
    [Fact]
    public void Runs_always_rejoin_to_the_input()
    {
        foreach (var input in new[]
                 {
                     "", "hello", "สวัสดี", "l;ylfu ้ำสสน ครับ 2024 :)", "hi🐈there",
                     "a\n\tb", "café", "日本", "  ", "ๆๆๆ!!!",
                 })
        {
            Assert.Equal(input, string.Concat(RunSplitter.Split(input).Select(r => r.Text)));
        }
    }

    /// Thai combining marks must stay attached to the run, not be split off as neutral.
    [Fact]
    public void Thai_combining_marks_stay_in_the_thai_run()
    {
        var runs = RunSplitter.Split("ครับ");
        var run = Assert.Single(runs);
        Assert.Equal(4, run.Text.EnumerateRunes().Count());
    }

    /// The six typographic substitutes are Latin despite sitting outside ASCII, so `don’t` is one
    /// run and `RunJudge` sees a word to judge.
    [Fact]
    public void A_curled_apostrophe_keeps_its_word_in_one_run()
    {
        var run = Assert.Single(RunSplitter.Split("don’t"));
        Assert.Equal(Script.Latin, run.Script);
        Assert.Equal("don’t", run.Text);
    }

    [Fact]
    public void Curled_text_still_rejoins_to_the_input_exactly()
    {
        var input = "“don’t” — สวัสดี 2024";
        Assert.Equal(input, string.Concat(RunSplitter.Split(input).Select(r => r.Text)));
    }
}
